Port twisted.words.protocols.irc to Python 3

rodrigc opened the pull request:
https://twistedmatrix.com/trac/ticket/6320

rodrigc committed:
Use @classmethod decorator

rodrigc committed:
Change long to int for Python 3

We do not need long here

rodrigc committed:
Use bytes for buffer

rodrigc committed:
reduce() is now obtained from the functools module in Python 3

rodrigc committed:
Use twisted.python.compat.unicode for Python 3

rodrigc committed:
Add missing mode

rodrigc committed:
Change map() to list comprehension

rodrigc committed:
Change iterkeys() to keys(), since iterkeys() does not exist in Python 3

rodrigc committed:
Use bytes for buffer

rodrigc committed:
Use byte strings

rodrigc committed:
Sort imports

rodrigc committed:
Use twisted.python.compat.unicode for Python 3

rodrigc committed:
Add a helper function to compare bytes buffer value to unicode string

rodrigc committed:
Use byte string

rodrigc committed:
Encode unicode string to bytes when we send, decode it from bytes when we receive

rodrigc committed:
Use byte string

rodrigc committed:
transport.value() is bytes, so use byte string when we split

rodrigc committed:
Remove unnecessary encode

rodrigc committed:
Compare buffer value

rodrigc committed:
Add missing modes which weren't getting parsed

rodrigc committed:
Decode buffer to unicode so we can process it as unicode

rodrigc committed:
Compare buffer value

rodrigc committed:
Change string.split(s) to s.split(), and string.lower(s) to s.lower()

string.split() and string.lower() static methods are gone in Python 3

rodrigc committed:
Compare buffer value which is always bytes with unicode string

rodrigc committed:
Sort imports

rodrigc committed:
Import twisted.compat.unicode for Python 3

rodrigc committed:
Sort imports

rodrigc committed:
Import twisted.python.compat for Python 3

rodrigc committed:
Use twisted.python.compat.itervalues() for Python 3

rodrigc committed:
Properly handle encoding/decoding for Unicode, motivated by Python 3

rodrigc committed:
Keep buffer as unicode

rodrigc committed:
Handle unicode properly for Python 3

Try to keep everything internally as unicode, and only
convert to bytes when sending over the transport.

rodrigc committed:
Use repr() to print out string in error message in case of Unicode decode error

rodrigc committed:
BADTEXT needs to be bytes, since it is used to test Unicode decode error handling

rodrigc committed:
Sort imports

rodrigc committed:
Compare byte string output to unicode string expected value

rodrigc committed:
Test that sending invalid bytes raises UnicodeError.

This is easier than trying to compare bad bytes in error message

rodrigc committed:
We need to return a list(), not a map iterator

rodrigc committed:
Decode response to unicode before processing the value

rodrigc committed:
Use unicode strings for name/password for authentication tests to fix test on Python 3

rodrigc committed:
Add topfile

glyph commented on src/twisted/words/test/test_irc.py:
```diff
@@ -9,13 +9,27 @@
+
+
+
+def assertEqualBufferValue(bufferValue, val):
```
As per https://twisted.readthedocs.io/en/twisted-16.4.1/core/development/policy/coding-standard.html#docstrings - "all methods, functions, classes, and modules must have docstrings".  So this should have a docstring :).  The docstring should describe its behavior, the type and purpose of `bufferVal` and `val` parameters.

rodrigc replied:
Fixed.

glyph commented on src/twisted/words/test/test_irc.py:
```diff
@@ -9,13 +9,27 @@
+    if isinstance(bufferValue, list):
+        if isinstance(val[0], unicode):
+            bufferValue = [buf.decode("utf8") for buf in bufferValue]
+    assert bufferValue == val, "%s != %s" % (bufferValue, val)
```
Assertions are generally frowned upon in Twisted, because [they aren't always run; `python -OO` disables them](https://docs.python.org/2/reference/simple_stmts.html#the-assert-statement).  This should be using `assertEqual`; which means this function should take a `TestCase` object as well.

rodrigc replied:
Fixed

glyph commented on src/twisted/words/test/test_irc.py:
```diff
@@ -1634,7 +1644,13 @@ def setUp(self):
-        self.assertEqual(self.f.getvalue(), s)
+        bufferValue = self.f.getvalue()
+        if isinstance(s, unicode):
+            # Convert bufferValue from bytes
```
Try to avoid using comments; instead, put this information into the docstring.  (i.e.: add a docstring explaining this method.)

rodrigc replied:
Fixed

glyph commented on src/twisted/words/test/test_irc.py:
```diff
@@ -9,13 +9,27 @@
 import operator
 import time
 
+from twisted.internet import protocol, task
```
There are lots and lots of string constants in this module and it _might_ be helpful to have them be a consistent type between 2/3; a `from __future__ import unicode_literals` might catch potential issues.

rodrigc replied:
@glyph I have addressed the coverage issues for ircsupport by adding more tests

rodrigc replied:
@glyph Where do you want `from __future__ import unicode_literals` to be added?
I ran an experiment and added it to:

```
twisted/words/im/ircsupport.py
twisted/words/protocols/irc.py
twisted/words/test/test_irc.py
twisted/words/test/test_irc_service.py
twisted/words/test/test_ircsupport.py
```

and some of the tests started breaking on Python 2, without breaking on Python 3.

glyph commented on src/twisted/words/test/test_irc.py:
```diff
@@ -792,14 +806,14 @@ def test_support_CHANMODES(self):
 
         self.assertEqual(
-            self._parseFeature('CHANMODES', ''),
+            self._parseFeature('CHANMODES', ',,,'),
```
Why is this change relevant to the 2->3 port?  These values really ought to be python-version independent...

rodrigc replied:
This is needed.  Without this change, the test will fail on Python 3 like this:

```
===============================================================================
[FAIL]
Traceback (most recent call last):
  File "/Users/crodrigues/twisted13/src/twisted/words/test/test_irc.py", line 821, in test_support_CHANMODES
    'noParam': ''})
  File "/Users/crodrigues/twisted13/src/twisted/trial/_synctest.py", line 425, in assertEqual
    super(_Assertions, self).assertEqual(first, second, msg)
  File "/usr/local/Cellar/python3/HEAD-42a10253d066_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/unittest/case.py", line 820, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/usr/local/Cellar/python3/HEAD-42a10253d066_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/unittest/case.py", line 1111, in assertDictEqual
    self.fail(self._formatMessage(msg, standardMsg))
twisted.trial.unittest.FailTest: {'addressModes': ''} != {'param': '', 'noParam': '', 'addressModes': '', 'setParam': ''}
- {'addressModes': ''}
+ {'addressModes': '', 'noParam': '', 'param': '', 'setParam': ''}

twisted.words.test.test_irc.ServerSupportedFeatureTests.test_support_CHANMODES
-------------------------------------------------------------------------------
```

The underlying _parseChanModesParam() function is using a combination of map, lambda, and dict which is triggering this difference in behavior between 2 and 3.

glyph reviewed (commented):
Thanks for doing this, @rodrigc.  It actually looks pretty good, except for the coverage issue.  Thankfully, the codecov report looks accurate this time: https://codecov.io/gh/twisted/twisted/compare/9fd621c0d425661bf23332cd1692ed256aa2e0b9...7eba1e677c22e36cf37141a98453cbf4b036335b#7372632F747769737465642F776F7264732F696D2F697263737570706F72742E7079-40 the issues are almost entirely in `ircsupport.py`.  One way forward would be to simply push those changes into a separate PR first, since aside from minor docstring issues, everything else here is ready to land.  ircsupport should be a pretty separate module, as well.

Can you address the docstring and coverage issues somehow and resubmit?  Thanks!

rodrigc committed:
Add some tests for ircsupport

rodrigc committed:
Make a subclass of unittest.TestCase.

Put assertEqualBufferValue() in this subclass.
Add some more tests for ircsupport.

markrwilliams commented on src/twisted/words/protocols/irc.py:
```diff
@@ -2620,9 +2622,17 @@ def connectionMade(self):
+
 
     def lineReceived(self, line):
+        if isinstance("", unicode) and isinstance(line, bytes):
```
`if isinstance("", unicode)` is a much better check than `if _PY3`

markrwilliams commented on src/twisted/words/protocols/irc.py:
```diff
@@ -2620,9 +2622,17 @@ def connectionMade(self):
     def lineReceived(self, line):
+        if isinstance("", unicode) and isinstance(line, bytes):
+            # decode bytes from transport to unicode
+            line = line.decode("utf-8")
```
See previous comment on encoding.

markrwilliams commented on src/twisted/words/protocols/irc.py:
```diff
@@ -2620,9 +2622,17 @@ def connectionMade(self):
-        basic.LineReceiver.dataReceived(self, data.replace('\r', ''))
+        data = data.replace('\r', '')
+        if isinstance(data, unicode):
+            data = data.encode("utf-8")
```
IRC is a binary protocol without a defined encoding.  That means on Pythons > 2 this can produce incorrect behavior, because messages can be in any encoding, or even multiple encodings.  Users of this API wouldn't be able to do much about that.

The protocol should not be implemented in terms of native strings any more than HTTP should be parsed in terms of native strings.  In fact, it'd be better to do this with HTTP, which _does_ have a defined encoding ([ISO 8859-1 for RFC 2616](https://www.w3.org/Protocols/rfc2616/rfc2616-sec2.html#sec2.2) and [ASCII for RFC 7230](https://tools.ietf.org/html/rfc7230#section-3.2.4)), but Twisted does not.  It shouldn't do so here.

If this _does_ do encoding and decoding, all the line length checks (e.g., [`_safeMaximumLineLength`](https://github.com/twisted/twisted/pull/534/files#diff-71c9155319411063da926db195c663d5R1684)) have to run after encoding, not before.  Otherwise a line of multi-byte UTF-8 code points will appear to be of an acceptable length but will actually be too long on the wire.

markrwilliams commented on src/twisted/words/protocols/irc.py:
```diff
@@ -2620,9 +2622,17 @@ def connectionMade(self):
 
     def dataReceived(self, data):
-        basic.LineReceiver.dataReceived(self, data.replace('\r', ''))
+        data = data.replace('\r', '')
```
I don't think this would work with a real TCP transport under Python 3, because the transport would call `dataReceived` with `bytes`, while `\r` and `''` are both native strings.

rodrigc replied:
Fixed.

markrwilliams commented on src/twisted/words/protocols/irc.py:
```diff
@@ -2898,7 +2908,7 @@ class DccChat(basic.LineReceiver, styles.Ephemeral):
     client = None
     remoteParty = None
-    buffer = ""
+    buffer = b""
```
If this is `bytes`, then `DccChat.lineReceived` should decode the line before sending it to `self.client.privmsg`, or `IRCClient` should do all of its work in `bytes` and rely on a higher level API to handle decoding.

markrwilliams reviewed (commented):
I really appreciate the work done here!

I strongly disagree with the assumption that `IRCClient` should work in terms of native strings, decoding bytes as UTF-8 via a hard-coded value.

IRC is a binary protocol and [lacks a defined encoding](https://tools.ietf.org/html/rfc2812#section-2.2). IRCv3 _may_ allow Unicode in nicknames and tags, and it looks like at least [some of the previous implementation's encoding](https://github.com/twisted/twisted/commit/733a1dd36cf426b009d076a8ee98114282b4bd27) was in the name of IRCv3 support.  But IRC 2.7 and earlier do not explicitly support any particular encoding.  (I mean to imply that IRCv3 support should live outside `IRCClient`).

I think `IRCClient` should deal only in byte strings, and require that users or a high-layer API (`DecodesIRCClient`?) decode and encode data.

rodrigc committed:
Add more tests for ircsupport

rodrigc committed:
IRCClient lacks a me() method.

rodrigc committed:
Add more tests for ircsupport

rodrigc committed:
Remove unused code

rodrigc committed:
Manipulate bytes in dataReceived()

rodrigc committed:
Decode bytes coming off wire to unicode, use unicode in application

rodrigc committed:
Fix whitespace

rodrigc committed:
Make sure we are sending bytes to basic.LineReceiver

rodrigc committed:
Convert comment to docstring

rodrigc commented:
In addition to the unit tests, I have been using this program to create an IRC server:

```
import sys

from twisted.cred import checkers, portal
from twisted.internet import reactor
from twisted.internet.endpoints import TCP4ServerEndpoint
from twisted.python import log
from twisted.words import service


ROOM = 'room'

if __name__ == '__main__':
    log.startLogging(sys.stdout)

    # Initialize the Cred authentication system used by the IRC server.
    realm = service.InMemoryWordsRealm('testrealm')
    realm.addGroup(service.Group(ROOM))
    user_db = checkers.InMemoryUsernamePasswordDatabaseDontUse()
    user_db.addUser("user1", "bar") 
    user_db.addUser("user2", "bar") 
    user_db.addUser("user3", "bar") 
    user_db.addUser("user4", "bar") 
    p = portal.Portal(realm, [user_db])
    p.registerChecker(checkers.AllowAnonymousAccess())

    # IRC server factory.
    ircfactory = service.IRCFactory(realm, p)

    # Connect a server to the TCP port 6667 endpoint and start listening.
    endpoint = TCP4ServerEndpoint(reactor, 6667)
    endpoint.listen(ircfactory)
    reactor.run()
```

I have run this server under Python 2 and Python 3.

I have then connected a few IRC clients  (ircII and Limechat) to this server and have
performed these operations:

/JOIN
/DCC CHAT
/DCC SEND
/WHO
/TOPIC

and they work.

nanonyme commented on src/twisted/words/protocols/irc.py:
```diff
@@ -405,6 +405,8 @@ def dataReceived(self, data):
         on and off was not required.)
         """
+        if isinstance(data, bytes):
+            data = data.decode("utf-8")
```
Naively decoding input as UTF-8 sounds like a bad idea without ability to set user-defined (probably per-channel) fallback encoding. Normal IRC clients tend to work so that they first try to decode as UTF-8 and if that fails, fallback to configurable legacy 8-bit encoding

glyph replied:
This is an important issue, but we could add encoding configurability in a follow-up PR.  Threading this through and providing the APIs to set it for each channel on all the various layers could be a separate issue.

rodrigc commented:
@glyph I just wanted to let you know that I did some additional testing with ircLogBot.py from http://twistedmatrix.com/documents/current/words/examples/

I did the test first with Python 2, then repeated the same test with Python 3.

I was able to connect the ircLogBot.py to `irc.freenode.net`, and was able to
connect the bot to a channel, and have it receive notifications.

Circlepuller commented:
I'm _really_ looking forward to this addition, as I don't believe there is really any decent IRC frameworks for Python 3.

rodrigc committed:
Port IRC modules to Python 3

rodrigc committed:
Fix whitespace

rodrigc committed:
Remove some assertions

rodrigc committed:
Add more test cases

rodrigc committed:
Add more tests

rodrigc committed:
Fix whitespace

rodrigc commented:
Moved to:
https://github.com/twisted/twisted/pull/561
